query_handling.py

```diff
@@ -45,6 +47,51 @@ def generate_queries(count_matrix, words: Dict[int, str], count: int, min_length
     return queries
 
 
+def generate_topic_queries(count_matrix,
+                           document_topic_matrix,
+                           words: Dict[int, str],
+                           count: int,
+                           min_length: int = 1,
+                           max_length: int = 4):
+    """
+    Generates queries for random topics based and samples from 1-4 documents from that topic
+    based on the topic distribution values.
+    It then adds the highest tfidf word to the query.
+    If that word is already in the query it takes the next one in the tf-idf list.
+    :param count_matrix: CountVectorization matrix
+    :param document_topic_matrix: dt matrix
+    :param words: words dictionary
+    :param count: number of queries wanted
+    :param min_length: min words per query (exact length is random)
+    :param max_length: max words per query (exact length is random)
+    :return: dictionary mapping document ids to queries
+    """
+    tfidf = TfidfTransformer()
+    tfidf_matrix = tfidf.fit_transform(count_matrix)
+    queries = {}
+    topic_count = document_topic_matrix.shape[1]
+    for i in tqdm(range(count)):
+        topic_id = random.randrange(0, topic_count)
+        query_length = random.randrange(min_length, max_length + 1)
+        query = []
+        topic_vec = tfidf_matrix.getcol(topic_id)
+        document_ids_in_topic = topic_vec.nonzero()[0]
+        normalized_topic_vector = normalize(topic_vec[document_ids_in_topic].toarray(), norm="l1", axis=0)
+        sampled_doc_ids = [np.random.choice(document_ids_in_topic, p=normalized_topic_vector[:, 0]) for x in
+                           range(0, query_length)]
+        for document in sampled_doc_ids:
+            index = 0
+            tf_idf_list = tfidf_matrix.getrow(document).toarray()[0].argsort()[-max_length:][::-1]
+            tf_idf_word = words[tf_idf_list[index]]
+            while tf_idf_word in query:
+                index += 1
+                tf_idf_word = words[tf_idf_list[index]]
+            query.append(tf_idf_word)
+        query = ' '.join(query)
+        queries[topic_id] = query
```

Comment:
Bliver en query for et topic ikke overskrevet hvis det topic blev tilfældigt valgt tidligere? Om count skal sættes til antal topics, hvis count er større, så vi kun får én query per topic maks, eller om der bare skal appendes values til dict keyen, kan jeg ikke sige.